### PR TITLE
MAINT: stats: move entropy and differential_entropy functions to more appropriate file

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue

gh-13631

#### What does this implement/fix?

In gh-13631, `differential_entropy` was added to the file where `entropy` was located. (makes sense)
`entropy` was located in `_distn_infrastructure`. (maybe not the best place)
Since they are listed as "summary statistics" in the documentation and `stats.py` is unmanageably long, makes some sense to put both of them with other summary statistics in `morestats.py`.

#### Additional information

I would suggest comparing the text of the functions that were deleted and those that are added to see that they are identical except that in `entropy` I've prepended `special.` to `entr` and `rel_entr` to avoid new imports.

### CODE REVIEW

**Review by tupui — APPROVED** [review]
LGTM. I am just wondering if we should not create a file for summary statistics instead. What do you think? If not, I will merge 😃

**Comment by mdhaber** [comment]
Yes and no. `morestats.py` is already getting long enough to slow down Spyder on my computer, so i don't love putting it there. But there are a few reasons i went with that. I'll think about it more.

What would you name the file? `summary_statistics.py`? And any new summary statistics would go there? What happens when _it_ gets long?

I suppose if I were to create a new file, I'd prefer for it to be just for entropy, e.g. `entropy.py`. i wanted to do that here, but there is some resistance to having lots of separate files like that, maybe because of import time?

**Comment by tupui** [comment]
Yes maybe `summary_statistics.py`. I trust your judgment here. You know better as you have this issue with everything to add in `stats`.

In the end, if `stats` becomes too big we might just have to split it. Personally, I would have put QMC outside, as a module. We can have this longer discussion on the mailing list or in another issue though.

**Comment by WarrenWeckesser** [comment]
> I suppose if I were to create a new file, I'd prefer for it to be just for entropy, e.g. `entropy.py`.

A separate file is fine with me, but make it explicit that the module is private by including the underscore prefix: `_entropy.py`

**Comment by mdhaber** [comment]
Usually I've done that (e.g. `_pagel`, `_linprog`, and others are mine) but I wondered why `morestats.py` still doesn't have an underscore.

**Comment by WarrenWeckesser** [comment]
> ... I wondered today why morestats.py still doesn't have an underscore.

Ancient history, lost in the mists of time, etc.  Adding the underscore is a potential backwards incompatible break.  The file has been there for a long time; who knows how much code is out there that explicitly imports from `scipy.stats.morestats`?

**Comment by mdhaber** [comment]
I occurs to me that a lot of functions have recently been imported then added to `__all__` in `stats.py` rather than being directly added to `__all__` in `__init__.py`. Is there any reason to do that unless the function is used in `stats.py`? 

Since `__init__.py` has the nice 
```python3
__all__ = [s for s in dir() if not s.startswith("_")]  # Remove dunders.
```
I think this adding to `__init__.py` is more maintainable (e.g. avoid all these merge conflicts as new functions are added.) 

I'm going to move the new functions (e.g. _hypotest.py, `_page_trend_test.py`) to `__init__.py` if that sounds OK.

File: scipy/stats/morestats.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+from __future__ import annotations
```
Comment (mdhaber) [inline scipy/stats/morestats.py:1]: Oops. Left this behind. I guess it can't hurt.

**Comment by WarrenWeckesser** [comment]
> I think this adding to `__init__.py` is more maintainable...

Yeah, that's what we should be doing.  We shouldn't add anything new to the `scipy.stats.stats` namespace.  New functions can be implemented in private modules, and then added to the `scipy.stats` namespace in `__init__.py`.

**Comment by mdhaber** [comment]
@WarrenWeckesser do you know if 
> Are they needed in stats.py if they are referenced in the docstrings of functions in stats.py?

**Comment by WarrenWeckesser** [comment]
For building the documentation, I don't think the names are needed in `stats.py`.  Given the way the docs are built, I'm pretty sure Sphinx will look for a reference such as `gmean` in the `scipy.stats` namespace.

**Comment by mdhaber** [comment]
OK. Since those are unrelated I won't fix here, but I'll do it separately soon. @tupui I think the PR as-is responds to your comment about putting these in a new file.

**Comment by tupui** [comment]
Yes, I am merging. Thanks everyone.

**Comment by rgommers** [comment]
Unfortunately `stats.distributions` is public API, see http://scipy.github.io/devdocs/reference/index.html#api-definition. It's not well-documented (e.g. doesn't have a submodule docstring), but I'm not sure we can get away with deprecation the whole module. Which also means you should remove functions from it. Can you put `stats.distributions.entropy` back? Can be an import rather than moving the implementation back.

**Comment by mdhaber** [comment]
Oops. Ok I'll do that and add a test to make sure it doesn't happen again.

**Pull request merged**